packages/next/src/server/app-render/app-render.tsx

```diff
@@ -1164,17 +1164,38 @@ async function renderToHTMLOrFlightImpl(
     let formState: null | any = null
     if (isActionRequest) {
       // For action requests, we handle them differently with a special render result.
-      const actionRequestResult = await handleAction({
-        req,
-        res,
-        ComponentMod,
-        serverModuleMap,
-        generateFlight: generateDynamicFlightRenderResult,
-        staticGenerationStore,
-        requestStore,
-        serverActions,
-        ctx,
-      })
+      const actionRequestResult =
+        await await ComponentMod.staticGenerationAsyncStorage.run(
+          {
+            ...staticGenerationStore,
+            // override isRender when handling actions
+            // if the action renders this should create
+            // a new static context where it sets back to true
+            isRender: false,
+          },
+          () => {
+            const currentStaticGenerationStore =
+              ComponentMod.staticGenerationAsyncStorage.getStore()
+
+            if (!currentStaticGenerationStore) {
+              throw new Error(
+                'Invariant missing static generation store in handleAction'
+              )
+            }
```

Comment:
The object we pass into `run` is the store, isn't it? So we could use this here and avoid the invariant error, right? (I guess this won't be relevant if we decide for the mutating approach?)